File: MapReduce_FastQ/src/main/java/PerBaseQual/FastQ_PerBaseQual_Mapper.java
package PerBaseQual;

import org.apache.hadoop.io.DoubleWritable;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.seqdoop.hadoop_bam.SequencedFragment;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 06.08.15.
 */
public class FastQ_PerBaseQual_Mapper extends org.apache.hadoop.mapreduce.Mapper<Object, SequencedFragment, IntWritable, DoubleWritable> {
    private static final int OFFSET = 33;

    @Override
    protected void map(Object key, SequencedFragment value, Context context) throws IOException, InterruptedException {
        final Text quality = value.getQuality();
        for (int i = 0; i < quality.getLength(); i++) {
            context.write(new IntWritable(i + 1), new DoubleWritable(getCorrespondingIntValue(quality.charAt(i))));
        }
    }

    private int getCorrespondingIntValue(int qualVal) {
        return qualVal - FastQ_PerBaseQual_Mapper.OFFSET;
    }

}

File: Spark_FastQ_Scala/src/main/scala/PerBaseQual/FastQ_PerBaseQual_Mapper.scala
package main.scala.PerBaseQual

import org.seqdoop.hadoop_bam.SequencedFragment
import scala.collection.mutable.ListBuffer
import org.apache.hadoop.io.Text

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 02.01.16.
  */
object FastQ_PerBaseQual_Mapper {
  private val OFFSET: Int = 33

  def flatMap(record: SequencedFragment): TraversableOnce[Pair[Int,Double]] = {
    val resList = new ListBuffer[Pair[Int,Double]]()
    val quality: Text = record.getQuality
    for(i <- 0 until quality.getLength) {
      resList.append(new Pair((i+1), getCorrespoindingIntValue(quality.charAt(i))))
    }
    resList.toTraversable
  }

  private def getCorrespoindingIntValue(qualVal: Int): Double = {
    (qualVal - OFFSET)
  }

}
